File: p.s.e.t.1/credit.c
/** 
 * credit.c
 *
 * a program that determines whether a provided credit card number is valid 
 * according to Luhn’s algorithm, per the below:
 *
 * $ ./credit
 * Number: 378282246310005
 * AMEX
 */

#include <stdio.h>
#include <stdlib.h>
#include <cs50.h>

// stores credit card number
long long int cc_number;

// function prototypes
int cal_checksum(void);
char val_identif(void);
void val_length(char company);

int main(void)
{
    int checksum; 
    char identifier = '\0';
    
    // prompt for user input
    printf("Number: ");
    cc_number = get_long_long();
    
    // calculate checksum
    checksum = cal_checksum();
    
    // validate checksum
    if (checksum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    
    // validate company's identifier
    identifier = val_identif();
    
    if (identifier == 'I')
    {
        printf("INVALID\n");
        return 0;
    }
    
    // validate number's lengtt and print result
    val_length(identifier);
    
    return 0;
}


/***** calculates checksum *****/
int cal_checksum(void)
{
    int mltpl = 0, check = 0;
    long long int numb = cc_number;
    
    // add digits to 'check' starting from rightmost
    while (numb != 0)
    {
        check += numb % 10;
        mltpl = ((numb /= 10) % 10) * 2;
        
        // add each digit separately 
        while (mltpl != 0)
        {
            check += mltpl % 10;
            check += (mltpl /= 10) % 10;
            mltpl /= 10;
        }
        
        // get rid of rightmost digit
        numb /= 10;
    }
    return check;
}


/***** validate company's identifier function *****/
char val_identif(void)
{
    char str_n[20], str_n2[2];
    int identif;
    char cmpn = '\0';
    
    // extract identirier
    sprintf(str_n, "%lli", cc_number);
    sprintf(str_n2, "%c%c", str_n[0], str_n[1]);
    identif = atoi(str_n2);
    
    // select company
    if (identif == 34 || identif == 37)
    {
        cmpn = 'A';
    }
    else if (identif > 50 && identif < 56)
    {
        cmpn = 'M';
    }
    else if (identif > 39 && identif < 49)
    {
        cmpn = 'V';
    }
    else
    {
        cmpn = 'I';
    }
    
    return cmpn;
}


/***** validate number's lengtt function *****/
void val_length(char company)
{
    int len = 0;
    
    // calculate length
    while (cc_number != 0)
    {
        len += 1;
        cc_number /= 10;
    }
        
    // validate length and print result
    if (company == 'A')
    {
        if (len == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (company == 'M')
    {
        if (len == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (company == 'V')
    {
        if (len == 13 || len == 16)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }    
}

File: p.s.e.t.6/credit.py
#  credit.py
# 
# a program that determines whether a provided credit card number is valid 
# according to Luhn’s algorithm, per the below:
#
# $ python credit.py
# Number: 378282246310005
# AMEX

import cs50
import math

def main():

    # prompt for user input
    print("Number: ", end="")
    number = cs50.get_int()
    str_number = str(number)
    list_number = list(str_number)
    
    # multiply every other digit
    multiply = []
    for i in range(len(list_number)//2):
        multiply.append(int(list_number[len(list_number)-(i*2)-2]) * 2)
    multiply.reverse()
    
    # sum those digits
    sum_multiply = 0
    for i in multiply:
        sum_multiply += sum_digits(i)
    
    # add to remaining digits
    sum_total = sum_multiply
    for i in range (math.ceil(len(list_number)/2)):
        sum_total += int(list_number[len(list_number)-1-i*2])

    # validate checksum, company's identifier, and number's length
    if sum_total % 10 != 0:
        print("INVALID")
    elif len(list_number) < 13 or len(list_number) == 14 or len(list_number) > 16:
        print("INVALID")
    elif list_number[0:1] == ['4']:
        print("VISA")
    elif list_number[0:2] == ['3', '4'] or list_number[0:2] == ['3', '5'] or list_number[0:2] == ['3', '6'] or list_number[0:2] == ['3', '7']:
        print("AMEX")
    elif list_number[0:2] == ['5', '1'] or list_number[0:2] == ['5', '2'] or list_number[0:2] == ['5', '3'] or list_number[0:2] == ['5', '4'] or list_number[0:2] == ['5', '5']:
        print("MASTERCARD")
    else:
        print("INVALID")

# sum digits of number
def sum_digits(n):
    s = 0 
    while n:
        s += n % 10
        n //= 10
    return s


if __name__ == "__main__":
    main()